### ISSUE TITLE
Compatibility across versions

### ISSUE THREAD

**Issue by tasleson** [pywbem/pywbem#511]
What's the projects stance with regards to compatibility?  Is the path to version 1 paved with incompatible code?

I'm seeing the following inconsistencies with running 0.7.0 vs. 0.10.0:

`pywbem.cim_http.Error -> pywbem.exception.Error or pywbem.Error`
`pywbem.cim_http.AuthError -> pywbem.exception.AuthError or pywbem.AuthError`

0.7.0 does not offer pywbem.Error & pywbem.AuthError

In addition CIMException on 0.7.0 uses [] to retrieve error code [0] and error msg [1].   Version 0.10.0 appears to use CIMException.status_code & CIMException.status_description.

There are likely other differences too, that I haven't stumbled upon.

**Comment by KSchopmeyer** [comment]
You are correct. There are some compatibility issues.  To be honest, we tried to keep compatibility but clearly that did not work out perfectly.  One thing we did not do was to keep a complete set of 7.0 tests just to insure comparability. We did significantly redo exceptions, partly to get them under control with the ssl but partly just for a cleaner interfaice but we ignored compatibility.  I am sure that there are some others but hopefully not many.  See the changes.rst for more details on our changes.

If you find others, please update us.

**Comment by andy-maier** [comment]
I'd like to add to what Karl said, by pointing to the [Deprecation policy](https://pywbem.readthedocs.io/en/latest/intro.html#deprecation-policy) section of the pywbem documentation. Even below 1.0.0, we are trying to stick to these rules as much as possible (i.e. 0.7 -> 0.8 -> 0.9 -> 0.10.0 are all considered new minor releases which can add new deprecations but are compatible). The text actually says "as backwards compatible as possible", leaving some wiggle room there. i think we should not allow any incompatibility in a minor release when we reach 1.0.0.

On the path to 1.0.0, my take is we try hard to stay compatible. Note that we have issue #497 open in order to discuss what we would like to change before we release a 1.0.0. Note that this is a wishlist, and we may end up not doing some of these due to compatibility issues. You can help us by commenting on that issue in case you have issues with the proposed items.

Having said all that, I think we can and should resolve the incompatibilities you listed:

* pywbem.cim_http.Error -> pywbem.Error: We can make that class available in the pywbem.cim_obj namespace, in addition. We likely cannot track its use and issue DeprecationWarning warnings.
* pywbem.cim_http.AuthError -> pywbem.AuthError: ditto
* 0.7.0 does not offer pywbem.Error & pywbem.AuthError: No issue, their addition is not a backwards compatibility issue.
* CIMException data access by list items (error code = [0], error msg = [1]): We can make that available again in a deprecated state (causing DeprecationWarning).

Please let us know any additional ones you find.

PS: I had started a compatibility checker a while ago. It is `tools/pycmp.py`, but it is probably not perfect.

**Comment by andy-maier** [comment]
Here is an update on the incompatibilities you found:

* pywbem.cim_http.Error -> pywbem.Error:

  That is actually a backwards compatibility issue. I will create a fix that imports that class again into `pywbem.cim_http`, for compatibility. I did not think someone would need the base class, but obviously that thought was incorrect because watching for the exception base class in a try block is a very common approach. I was probably just confused when I removed it.

* pywbem.cim_http.AuthError -> pywbem.AuthError:

  The class is actually still available in its old namespace, in addition to the new one:

  ```
  $ python -c "import pywbem; print(pywbem.cim_http.AuthError, pywbem.AuthError)"
  (<class 'pywbem.exceptions.AuthError'>, <class 'pywbem.exceptions.AuthError'>)
  ```

  If you still think there is a backwards compatibility issue of some sort, please explain.

* 0.7.0 does not offer pywbem.Error & pywbem.AuthError:

  True, but that is not a compatibility issue, because their addition does not break backwards compatibility.

* CIMException data access by list items (error code = [0], error msg = [1]):

  I verified that access by index (and by slice) still works, for compatibility:

  ```
>>> import pywbem
>>> e=pywbem.CIMError(12,"abc")
>>> e
CIMError(12, 'abc')
>>> e[0]
12
>>> e[1]
'abc'
>>> e[0:2]
(12, 'abc')
  ```

  The new properties have been added and thus are not a backwards compatibility issue.

  Pywbem inherits this access-by-index capability of `CIMError` from the built-in `Exception` class. I did not find this behavior of the built-in `Exception` class described anywhere in the Python docs. In the pywbem docs, we describe that the `args` attribute is set to the tuple(status_code, status_description), so it is clear that e.g. `exc.args[0]` is valid. However, we don't describe that `exc[0]` is also valid. That is the inherited capability that not even Python describes.

  Given that this behavior of the `Exception` class has been there all the time, and giventhat Python does not describe it, I think we should describe it in the pywbem docs. I'll include that in the upcoming fix.

**Comment by tasleson** [comment]
@andy-maier This is what I'm seeing:

```python
Python 3.5.1 (default, Aug  9 2016, 15:35:51) 
[GCC 6.1.1 20160621 (Red Hat 6.1.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pywbem
>>> e = pywbem.CIMError(777, "Some error")
>>> e
CIMError(777, 'Some error')
>>> e[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'CIMError' object does not support indexing
>>> pywbem.__version__
'0.10.0.dev0'
```
This looks to be an issue between py2 and py3.  As we have no other way to retrieve the status code and status message without using [], it would be good to have this work on newer versions for both py2 & py3.

**Comment by tasleson** [comment]
@andy-maier 

You are correct `pywbem.cim_http.AuthError` is available for both 0.7.0 and 0.10.0.

**Comment by tasleson** [comment]
@KSchopmeyer @andy-maier

Thank you for all your efforts on improving and updating pywbem.  It's a very useful piece of software!

A little background in hopes of clarifying our needs and concerns...

Our [project](https://github.com/libstorage/libstoragemgmt) that is using pywbem is trying to support py2 & py3 across linux distributions that will span many years (2010 to present).  Having to test our code against py2 and py3 is proving difficult because we are running against storage hardware that takes hours to complete.  If we have to repeat the tests with pywbem 0.7.0/0.8.0/0.9.0/0.10.0 etc. the test matrix becomes impossible.  I cannot stress the importance of API stability enough for what we are working on.  The hassle of testing py2 & py3 is getting to the point that we may just start re-writing the python code in something else so that we can just test it once.  This coupled with libraries that we depend on with subtle API differences is maddening.  I know there is an incredible urge to "clean" things up, make better, prettier when working with a library, but please be very aware that by doing so you may alienate the very people that you're trying to help.

**Comment by andy-maier** [comment]
On accessing the exception object by index/slice: I am also just learning about this. As it seems, Python 3 has removed that ability. I have adjusted the new unit testcase for exceptions to verify that the access works when running on python 2, and to verify that it fails on python 3.

**Comment by andy-maier** [comment]
Again on accessing the exception object by index/slice: With the current implementations of CIMError in pywbem, the only compatible way to access the status code across all versions of pywbem and across both python 2 and 3 is to access `exc.args[0]`. The new unit test case for exceptions tests this among many other tests. PR #513 verifies that for pywbem 0.9, and PR #514 for pywbem 0.10.

Please let me know if changing to that way of accessing it is acceptable.

**Comment by tasleson** [comment]
@andy-maier Yes, I should be able to use `exc.args[n]`, thanks!

**Comment by andy-maier** [comment]
Ok, then the PRs I created address everything that came up so far, IMO.
Feel free to review them (or just one of them, they have the same content) and give +1 or -1.

**Comment by andy-maier** [comment]
I just verified that in pywbem 0.8, the `Error` class was still defined in the `cim_http` module, so that compatibility issue does not exist in 0.8.x.

**Comment by andy-maier** [comment]
I suggest to close this issue here, because all reported issues that ended up being compatibility issues relative to 0.7 have fixes in place, in all pywbem releases 0.8, 0.9, and 0.10-dev.

**Comment by tasleson** [comment]
@andy-maier Thanks!
